File: backend/app/models/training_plan.py
from pydantic import BaseModel, Field
from typing import List, Optional, Literal
import uuid
from datetime import date, datetime # Import date and datetime

class WeeklySummary(BaseModel):
    week_number: int = Field(..., description="Week number (e.g., 1, 2, ...)")
    summary: str = Field(..., description="Brief description of the week's focus and key workouts (e.g., 'Focus on base mileage. Long run: 8 miles. Include 1 tempo run.')")
    estimated_weekly_mileage: str | None = Field(None, description="Estimated total mileage for the week (e.g., '20-25 miles')")
    # Optional: Could add fields for total mileage, key workouts list, etc.

class TrainingPlanOutline(BaseModel):
    race_name: str
    race_distance: str
    total_weeks: int
    weeks: List[WeeklySummary] = Field(..., description="List of weekly training summaries leading up to the race")
    notes: List[str] | None = Field(None, description="Optional additional notes or disclaimers from the AI")

# --- Detailed Daily Plan Pydantic Models ---

class DailyWorkout(BaseModel):
    date: str # ISO 8601 Format: "YYYY-MM-DD"
    day_of_week: Literal['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
    workout_type: Literal['Easy Run', 'Tempo Run', 'Intervals', 'Speed Work', 'Long Run', 'Rest', 'Cross-Training', 'Strength', 'Race Pace', 'Warm-up', 'Cool-down', 'Other']
    description: str # e.g., "5 miles at conversational pace"
    distance: Optional[str] = None # e.g., "5 miles"
    duration: Optional[str] = None # e.g., "45 minutes"
    intensity: Optional[str] = None # e.g., "Easy", "Tempo"
    notes: Optional[List[str]] = None
    status: Literal['pending', 'completed', 'skipped'] = 'pending' # Default to pending
    google_event_id: Optional[str] = None # ID of the event created in Google Calendar

class DetailedWeek(BaseModel):
    week_number: int
    start_date: str # "YYYY-MM-DD" of the Monday
    end_date: str # "YYYY-MM-DD" of the Sunday
    days: List[DailyWorkout] # Array of 7 DailyWorkout objects
    weekly_focus: Optional[str] = None
    estimated_weekly_mileage: Optional[str] = None

class DetailedTrainingPlan(BaseModel):
    plan_id: str = Field(default_factory=lambda: str(uuid.uuid4())) # Generate UUID for new plans
    user_id: str # Needs to be set when creating/saving
    race_name: str
    race_distance: str
    race_date: str # "YYYY-MM-DD"
    goal_time: Optional[str] = None # <-- ADDED: User's desired finish time (e.g., "1:45:00")
    plan_start_date: str # "YYYY-MM-DD" - First Monday of the plan
    total_weeks: int
    weeks: List[DetailedWeek]
    overall_notes: Optional[List[str]] = None
    personalization_details: Optional[dict] = None # Keep flexible for now
    generated_at: str = Field(default_factory=lambda: datetime.now().isoformat()) # Timestamp
    plan_version: str = "v2.0-daily" # Version indicator

# --- Model for Updating Daily Workout Status --- 

class DailyWorkoutStatusUpdate(BaseModel):
    status: Literal['pending', 'completed', 'skipped'] 

# --- Models for Workout Analysis --- 

class PlanContextForAnalysis(BaseModel):
    """Optional context about the overall plan for the AI."""
    race_name: Optional[str] = None
    race_distance: Optional[str] = None
    goal_time: Optional[str] = None
    plan_total_weeks: Optional[int] = None
    week_number: Optional[int] = None # The week number this workout belongs to
    pr_used: Optional[str] = None

class WorkoutAnalysisRequest(BaseModel):
    """Request body for the workout analysis endpoint."""
    race_id: uuid.UUID # Although redundant with URL, good for validation
    workout: DailyWorkout # The specific workout object, including user notes
    plan_context: Optional[PlanContextForAnalysis] = None

class WorkoutAnalysisResponse(BaseModel):
    """Response body containing the AI feedback."""
    feedback: str 

File: frontend/src/types/training_plan.ts
export interface WeeklySummary {
    week_number: number;
    summary: string;
    estimated_weekly_mileage?: string | null;
}

export interface TrainingPlanOutline {
    race_name: string;
    race_distance: string;
    total_weeks: number;
    weeks: WeeklySummary[];
    notes?: string[] | null;
}

// --- Detailed Daily Plan Structures ---

// Represents a single day's activity within the training plan
export interface DailyWorkout {
  date: string; // ISO 8601 Format: "YYYY-MM-DD"
  day_of_week: 'Monday' | 'Tuesday' | 'Wednesday' | 'Thursday' | 'Friday' | 'Saturday' | 'Sunday';
  workout_type: 'Easy Run' | 'Tempo Run' | 'Intervals' | 'Speed Work' | 'Long Run' | 'Rest' | 'Cross-Training' | 'Strength' | 'Race Pace' | 'Warm-up' | 'Cool-down' | 'Other';
  description: string; // Primary instruction, e.g., "5 miles at conversational pace", "Rest day", "30 min cycling"
  distance?: string; // Optional: e.g., "5 miles", "10 km"
  duration?: string; // Optional: e.g., "45 minutes", "1:30:00"
  intensity?: string; // Optional: e.g., "Easy", "Tempo", "Hard", "Conversational", "HR Zone 2"
  notes?: string[]; // Optional: Array of additional tips, instructions, or variations
  status: 'pending' | 'completed' | 'skipped'; // Status for user tracking
}

// Represents a single week within the training plan
export interface DetailedWeek {
  week_number: number; // Relative to the plan (1 to total_weeks)
  start_date: string; // "YYYY-MM-DD" of the Monday of this week
  end_date: string; // "YYYY-MM-DD" of the Sunday of this week
  days: DailyWorkout[]; // Array of 7 DailyWorkout objects for the week
  weekly_focus?: string; // Optional: A short summary of the week's goal
  estimated_weekly_mileage?: string; // e.g., "35-40 miles"
}

// Represents the overall detailed training plan
export interface DetailedTrainingPlan {
  plan_id: string; // Unique identifier for the plan instance
  user_id: string; // Identifier for the user this plan belongs to
  race_name: string;
  race_distance: string;
  race_date: string; // "YYYY-MM-DD"
  goal_time?: string; // Optional user goal
  plan_start_date: string; // "YYYY-MM-DD" - First Monday of the plan
  total_weeks: number;
  weeks: DetailedWeek[]; // The core daily schedule
  overall_notes?: string[]; // General notes from the coach/system for the whole plan
  personalization_details?: {
      pr_used?: string; // e.g., "5K PR: 25:30"
      adjustments_made?: string[]; // e.g., ["Reduced initial long run based on recent activity"]
  };
  generated_at: string; // ISO timestamp
  plan_version: string; // e.g., "v2.0-daily"
} 